content/docs/reference/how.md

```diff
@@ -7,44 +7,89 @@ menu:
+Next, we run `pulumi up`. Since this program is written in JavaScript, the Pulumi CLI launches the Node language host and requests that it execute the program. When the first `aws.s3.Bucket` object is constructed, the language host sends as _resource registration_ request to the deployment engine and then continues executing the program. This is subtle, but important: _When the call to `new aws.s3.Bucket` returns, it does not mean that the actual S3 bucket has been created in AWS_, it just means the language host has expressed that this bucket is part of the desired state of your infrastructure.  The language host continues to execute your program concurrently with the engine processing this request.
+
+In this case, since the last deployed state has no resources, the engine determines that it needs to create the `media-bucket` resource. It uses the AWS resource plugin in order to create the resource and the AWS resource plugin uses the AWS SDK in order to go create it.  Note that the engine does not talk directly to AWS, instead it just asks the AWS Resource Plugin to create a Bucket. As new resource types are added, you can simply update the version of a resource provider to gain access to these new resources without having to update the CLI itself. When the operation to create this bucket is complete, the engine records information about the newly created resource in its state file.
+
+As the engine was creating the `media-bucket` bucket, the language host continued to execute the Pulumi program. This caused another resource registration to be generated (for `content-bucket`). Since there is no dependency between these two buckets, the engine is able to process that request in parallel with the creation of `media-bucket`.
 
-The engine consults the last deployed state on pulumi.com, and determines that these resources do not already exist. So, the engine calls the AWS resource provider, requesting that it creates the two AWS S3 buckets. Once the operations succeed, this state is written to the last-deployed checkpoint. So, the resource list will be similar to the following:
+After both operations have completed, the language host exists (as the program has finished running) and then engine and resource providers shutdown. The state for mystack now looks like the following:
 
 ```
 stack mystack
    - aws.s3.Bucket "media-bucket653a4"
    - aws.s3.Bucket "content-bucket125ce"
 ```
 
-Now, suppose we rename `content-bucket` to `app-bucket`:
+Note the extra suffixes on the end of these bucket names. This is due to a process called [auto-naming]({{< relref "/docs/reference/programming-model.md#autonaming" >}}), which Pulumi uses by default in order to allow you to deploy multiple copies of your infrastructure without creating name collisions for resources. This behavior can be disabled if desired.
```

Comment:
> This behavior can be disabled if desired.

Is this documented anywhere on the site that we could link to from here?

Reply:
the section on auto-naming in the programming model page actually explains how to do this.